File: backend/src/services/blockchainService.ts
import { ethers } from 'ethers';
import * as crypto from 'crypto';

// Blockchain configuration
const PRIVATE_KEY = process.env.BLOCKCHAIN_PRIVATE_KEY || '0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80'; // Default Hardhat Account 0
const PROVIDER_URL = process.env.BLOCKCHAIN_PROVIDER_URL || 'http://127.0.0.1:8545';
const CONTRACT_ADDRESS = process.env.CONTRACT_ADDRESS || '0x5FbDB2315678afecb367f032d93F642f64180aa3';
const NETWORK_NAME = process.env.BLOCKCHAIN_NETWORK_NAME || 'Hardhat Localhost (Chain ID: 31337)';

const contractAbi = [
  "function recordEvent(string memory batchId, string memory eventType, string memory actorId, string memory dataHash, string memory previousEventHash) public",
  "function getEvents(string memory batchId) public view returns (tuple(string batchId, string eventType, string actorId, string dataHash, string previousEventHash, uint256 timestamp)[])",
  "function recordHarvesterVerification(string memory verificationId, string memory harvesterId, string memory recordHash, string memory status) public",
  "function getHarvesterVerification(string memory verificationId) public view returns (tuple(string verificationId, string harvesterId, string recordHash, string status, uint256 timestamp))",
  "function isHarvesterVerified(string memory verificationId) public view returns (bool)",
  "event ProvenanceRecorded(string indexed batchId, string eventType, string actorId, string dataHash, string previousEventHash, uint256 timestamp)",
  "event HarvesterVerified(string indexed verificationId, string indexed harvesterId, string recordHash, string status, uint256 timestamp)"
];

class BlockchainService {
  private provider: ethers.JsonRpcProvider;
  private wallet: ethers.Wallet;
  private contract: ethers.Contract;

  constructor() {
    this.provider = new ethers.JsonRpcProvider(PROVIDER_URL);
    this.wallet = new ethers.Wallet(PRIVATE_KEY, this.provider);
    this.contract = new ethers.Contract(CONTRACT_ADDRESS, contractAbi, this.wallet);
  }

  /**
   * Check if blockchain node is reachable
   */
  async checkConnection(): Promise<boolean> {
    try {
      await this.provider.getBlockNumber();
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Record Harvester Verification on blockchain
   */
  async recordHarvesterVerificationOnChain(
    verificationId: string,
    harvesterId: string,
    recordHash: string,
    status: string = 'VERIFIED'
  ): Promise<{
    success: boolean;
    txHash?: string;
    blockNumber?: number;
    network: string;
    isOffline?: boolean;
    error?: string;
  }> {
    try {
      const isOnline = await this.checkConnection();
      if (!isOnline) {
        console.warn(`[Blockchain] Node at ${PROVIDER_URL} is unreachable. Recording tamper-evident hash off-chain.`);
        return {
          success: false,
          isOffline: true,
          network: `${NETWORK_NAME} (Offline - Ledger Ready)`,
          error: 'Blockchain node is currently unreachable. Verification hash recorded with tamper-evidence for node sync.'
        };
      }

      // Execute transaction
      const tx = await (this.contract as any).recordHarvesterVerification(
        verificationId,
        harvesterId,
        recordHash,
        status
      );

      const receipt = await tx.wait();

      return {
        success: true,
        txHash: receipt.hash,
        blockNumber: receipt.blockNumber,
        network: NETWORK_NAME
      };
    } catch (error: any) {
      console.error('[Blockchain] Error recording harvester verification:', error?.message || error);
      return {
        success: false,
        network: NETWORK_NAME,
        error: error?.message || String(error)
      };
    }
  }

  /**
   * Query Harvester Verification from blockchain
   */
  async getHarvesterVerificationOnChain(verificationId: string): Promise<{
    found: boolean;
    record?: {
      verificationId: string;
      harvesterId: string;
      recordHash: string;
      status: string;
      timestamp: number;
    };
    error?: string;
  }> {
    try {
      const isOnline = await this.checkConnection();
      if (!isOnline) {
        return { found: false, error: 'Blockchain node is offline' };
      }

      const raw = await (this.contract as any).getHarvesterVerification(verificationId);
      if (!raw || !raw[0] || raw[0] === '') {
        return { found: false };
      }

      return {
        found: true,
        record: {
          verificationId: raw[0],
          harvesterId: raw[1],
          recordHash: raw[2],
          status: raw[3],
          timestamp: Number(raw[4])
        }
      };
    } catch (error: any) {
      return { found: false, error: error?.message || String(error) };
    }
  }

  /**
   * Compute cryptographic SHA-256 hash of canonical verification payload
   */
  computeVerificationHash(payload: Record<string, any>): string {
    const sortedKeys = Object.keys(payload).sort();
    const canonicalObj: Record<string, any> = {};
    for (const key of sortedKeys) {
      canonicalObj[key] = payload[key];
    }
    const jsonString = JSON.stringify(canonicalObj);
    return crypto.createHash('sha256').update(jsonString).digest('hex');
  }

  get contractAddress(): string {
    return CONTRACT_ADDRESS;
  }

  get networkName(): string {
    return NETWORK_NAME;
  }
}

export const blockchainService = new BlockchainService();

File: backend/services/blockchain_service.py
"""
Blockchain provenance service using Web3.py.
Interacts with the HoneyChainProvenance smart contract with offline-safe tamper-evident hashing.
"""
from __future__ import annotations

import os
import json
import hashlib
import logging
from typing import Dict, Any, Optional

try:
    from web3 import Web3
except ImportError:
    Web3 = None

logger = logging.getLogger("BlockchainService")

# Configuration
RPC_URL = os.getenv("BLOCKCHAIN_PROVIDER_URL", os.getenv("POLYGON_AMOY_RPC_URL", "http://127.0.0.1:8545"))
PRIVATE_KEY = os.getenv("BLOCKCHAIN_PRIVATE_KEY", "")
CONTRACT_ADDRESS = os.getenv("CONTRACT_ADDRESS", "")
CHAIN_ID = os.getenv("BLOCKCHAIN_CHAIN_ID", "")
NETWORK_NAME = os.getenv("BLOCKCHAIN_NETWORK_NAME", "Polygon Amoy (Chain ID: 80002)" if "amoy" in RPC_URL else "Hardhat Localhost (Chain ID: 31337)")

CONTRACT_ABI = [
    {
        "inputs": [
            {"internalType": "string", "name": "batchId", "type": "string"},
            {"internalType": "string", "name": "eventType", "type": "string"},
            {"internalType": "string", "name": "actorId", "type": "string"},
            {"internalType": "string", "name": "dataHash", "type": "string"},
            {"internalType": "string", "name": "previousEventHash", "type": "string"},
        ],
        "name": "recordEvent",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function",
    },
    {
        "inputs": [{"internalType": "string", "name": "batchId", "type": "string"}],
        "name": "getEvents",
        "outputs": [
            {
                "components": [
                    {"internalType": "string", "name": "batchId", "type": "string"},
                    {"internalType": "string", "name": "eventType", "type": "string"},
                    {"internalType": "string", "name": "actorId", "type": "string"},
                    {"internalType": "string", "name": "dataHash", "type": "string"},
                    {"internalType": "string", "name": "previousEventHash", "type": "string"},
                    {"internalType": "uint256", "name": "timestamp", "type": "uint256"},
                ],
                "internalType": "struct HoneyChainProvenance.BatchEvent[]",
                "name": "",
                "type": "tuple[]",
            }
        ],
        "stateMutability": "view",
        "type": "function",
    },
]


class BlockchainService:
    def __init__(self):
        self.rpc_url = RPC_URL
        self.contract_address = CONTRACT_ADDRESS
        self.network = NETWORK_NAME
        self.w3 = None
        self.contract = None
        self.account = None
        self.chain_id = None

        if Web3 and PRIVATE_KEY and CONTRACT_ADDRESS:
            try:
                self.w3 = Web3(Web3.HTTPProvider(self.rpc_url, request_kwargs={"timeout": 2.0}))
                if not self.w3.is_connected():
                    raise ConnectionError(f"Cannot reach blockchain RPC at {self.rpc_url}")
                self.chain_id = self.w3.eth.chain_id
                if CHAIN_ID and self.chain_id != int(CHAIN_ID):
                    raise ValueError(f"Configured chain ID {CHAIN_ID} does not match RPC chain ID {self.chain_id}")
                self.account = self.w3.eth.account.from_key(PRIVATE_KEY)
                checksum_address = Web3.to_checksum_address(self.contract_address)
                if self.w3.eth.get_code(checksum_address) in (b"", b"0x"):
                    raise ValueError(f"No contract bytecode found at {checksum_address} on chain {self.chain_id}")
                self.contract = self.w3.eth.contract(address=checksum_address, abi=CONTRACT_ABI)
                logger.info(f"Connected to blockchain node at {self.rpc_url} (chain ID {self.chain_id})")
            except Exception as e:
                self.w3 = None
                self.contract = None
                self.account = None
                logger.warning(f"Blockchain initialization failed: {e}")
        else:
            logger.info("Blockchain running in offline-safe tamper-evident hashing mode.")

    def is_connected(self) -> bool:
        try:
            return self.w3 is not None and self.w3.is_connected()
        except Exception:
            return False

    def compute_hash(self, data: Any) -> str:
        serialized = json.dumps(data, sort_keys=True, default=str)
        return hashlib.sha256(serialized.encode("utf-8")).hexdigest()

    def record_batch_event(
        self,
        batch_id: str,
        event_type: str,
        actor_id: str,
        payload: Dict[str, Any],
        previous_event_hash: str = "",
    ) -> Dict[str, Any]:
        data_hash = self.compute_hash(payload)

        if not self.is_connected():
            logger.info(f"[Blockchain Offline] Recording tamper-evident hash for {batch_id} ({event_type}): {data_hash[:16]}...")
            return {
                "success": False,
                "status": "PENDING",
                "data_hash": data_hash,
                "tx_hash": None,
                "block_number": None,
                "network": f"{self.network} (Offline - Ledger Ready)",
                "error": "Blockchain node offline. Tamper-evident hash logged for reconciliation.",
            }

        try:
            tx = self.contract.functions.recordEvent(
                batch_id,
                event_type,
                actor_id,
                data_hash,
                previous_event_hash,
            ).build_transaction({
                "from": self.account.address,
                "nonce": self.w3.eth.get_transaction_count(self.account.address),
                "gas": 300000,
                "gasPrice": self.w3.eth.gas_price,
            })

            signed_tx = self.w3.eth.account.sign_transaction(tx, private_key=PRIVATE_KEY)
            raw_transaction = getattr(signed_tx, "raw_transaction", getattr(signed_tx, "rawTransaction", None))
            if raw_transaction is None:
                raise RuntimeError("Web3 returned a signed transaction without raw transaction bytes")
            tx_hash = self.w3.eth.send_raw_transaction(raw_transaction)
            receipt = self.w3.eth.wait_for_transaction_receipt(tx_hash, timeout=10)

            tx_hex = self.w3.to_hex(tx_hash)
            logger.info(f"[Blockchain] Event {event_type} committed on-chain: {tx_hex}")
            return {
                "success": True,
                "status": "CONFIRMED",
                "data_hash": data_hash,
                "tx_hash": tx_hex,
                "block_number": receipt.blockNumber,
                "network": self.network,
                "error": None,
            }
        except Exception as err:
            logger.error(f"[Blockchain Error] Failed to submit transaction: {err}")
            return {
                "success": False,
                "status": "FAILED",
                "data_hash": data_hash,
                "tx_hash": None,
                "block_number": None,
                "network": self.network,
                "error": str(err),
            }

    def get_batch_events(self, batch_id: str) -> Dict[str, Any]:
        """Read the contract's provenance history; database records remain the API cache."""
        if not self.is_connected():
            return {"success": False, "events": [], "error": "Blockchain node or deployed contract is not configured."}
        try:
            events = self.contract.functions.getEvents(batch_id).call()
            return {
                "success": True,
                "events": [
                    {
                        "batchId": event[0], "eventType": event[1], "actorId": event[2],
                        "dataHash": event[3], "previousEventHash": event[4], "timestamp": int(event[5]),
                    }
                    for event in events
                ],
                "error": None,
            }
        except Exception as err:
            logger.error(f"[Blockchain Error] Failed to read events for {batch_id}: {err}")
            return {"success": False, "events": [], "error": str(err)}


blockchain_service = BlockchainService()
